### PR TITLE
Add filter_na function

### PR DESCRIPTION
**Pull request by jrnold**
This function filters rows that have any `NA`, or all `NA` values. It does so by generating the expressions to pass to `filter_()` which means that it will work for any backend supported by dplyr.

This is somewhat related to #1367.
There are also a few stack overlow questions asking how to do this: [here](http://stackoverflow.com/questions/26665319/removing-na-in-dplyr-pipe), [here](http://stackoverflow.com/questions/28857653/removing-na-observations-with-dplyrfilter), and [here](http://stackoverflow.com/questions/22353633/filter-for-complete-cases-in-data-frame-using-dplyr-case-wise-deletion). The answers given are to use `na.omit` or `complete.cases` which don't extend to other backends supported by dplyr.

### CODE REVIEW

File: tests/testthat/test-filter-na.R [hunk 1]
```diff
@@ -0,0 +1,24 @@
+context("filter_na")
+
+x <- tibble(a = c(1, 2, NA, NA),
```
Comment (krlmlr) [inline tests/testthat/test-filter-na.R]: Using tribble() for construction might make the tests easier to understand.

The first test would be stronger if c[4] was non-NA.

File: R/filter-na.r [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  }
+  vars <- select_vars_(names(.data), dots)
+  # construct expression for filter
+  # !is.na(var1), !is.na(var2), ...
+  args <- lapply(vars, function(i) call("!", call("is.na", as.name(i))))
+  if (.all) {
+    # !is.na(var1) | !is.na(var2) | ...
+    if (length(args) > 1) {
+      args <- Reduce(function(x, y) call("|", x, y),
+                     args[2:length(args)], args[[1]])
```
Comment (krlmlr) [inline R/filter-na.r]: Does `args[-1]` work here?

File: DESCRIPTION [hunk 3]
```diff
@@ -11,9 +11,9 @@ Authors@R: c(
     )
 URL: https://github.com/hadley/dplyr
 BugReports: https://github.com/hadley/dplyr/issues
-Depends: 
+Depends:
```
Comment (krlmlr) [inline DESCRIPTION]: Could you please undo the unnecessary changes in this file?

File: R/filter-na.r [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  if (length(dots) == 0) {
+    dots <- list(~ everything())
+  }
+  vars <- select_vars_(names(.data), dots)
+  # construct expression for filter
+  # !is.na(var1), !is.na(var2), ...
+  args <- lapply(vars, function(i) call("!", call("is.na", as.name(i))))
+  if (.all) {
+    # !is.na(var1) | !is.na(var2) | ...
+    if (length(args) > 1) {
```
Comment (krlmlr) [inline R/filter-na.r]: Do we need the special case here?

File: R/filter-na.r [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+#' @param ... Comma separated list of unquoted expressions. Select variables
+#'   which to test for missing values using the same semantics as
+#'   \code{\link[dplyr]{select}}.
+#' @param .all If \code{TRUE}, then remove rows in which all columns are
+#'   missing, else select rows in which any column has a missing value.
+#' @param .dots use \code{filter_na_()} to do standard evaluation. See
+#'   \code{vignette("nse")} for details.
+#' @return An object of the same class as \code{.data}.
+#' @export
+filter_na <- function(.data, ..., .all = FALSE) {
```
Comment (krlmlr) [inline R/filter-na.r]: Can you give a practical use case for `.all = FALSE`? How do I show only the rows that have NA in at least one column?

**Comment by krlmlr** [comment]
Thinking about it -- maybe it's a better fit for tidyr?

**Comment by hadley** [comment]
`filter_na()` feels to specific to be dplyr to me. But maybe dplyr could have `filter_if()` + `filter_all()` that worked similarly to the summarise and mutate equivalents?

**Comment by krlmlr** [comment]
So the syntax would be `data %>% filter_all(funs(!is.na(.)))` ?

**Comment by hadley** [comment]
Yeah or `data %>% filter_all(not_na)` if we also provided a helper.

File: R/filter-na.r (hunk 5 above)
Comment (jrnold) [inline R/filter-na.r]: It is removing all the rows with an `NA` in at least one column, as in `na.omit` or `complete.cases`

File: R/filter-na.r (hunk 4 above)
Comment (jrnold) [inline R/filter-na.r]: Yes, or I need to use `init = TRUE`, otherwise `Reduce` generates `NULL | ...`. `purrrr::reduce` is much nicer :-)

File: R/filter-na.r (hunk 2 above)
Comment (jrnold) [inline R/filter-na.r]: yeah, my bad.

**Comment by jrnold** [comment]
Are a `filter_if` and `filter_any`  in the works?

**Comment by jrnold** [comment]
The other feature that I had added in [my original version ](https://github.com/jrnold/rubbish/blob/master/R/filter_na.R) that I didn't include here was the option of using `is.finite` instead of `!is.na` for local tables and numeric variables.

File: R/filter-na.r (hunk 5 above)
Comment (jrnold) [inline R/filter-na.r]: And for the case of `NA` in at least one column, I hadn't thought about that.

**Comment by krlmlr** [comment]
I'm not aware of any implementation attempts. Would it be difficult to change this PR to implement `filter_if()` and `filter_any()`?

**Comment by jrnold** [comment]
I think that the pull request would likely have to be separate.  I didn't write this function with too much of a view for generality, focusing more on having a usable function for a common case. So I'd more about what the general set of functions would look like and rewrite completely to account for that.   

I'm not sure whether it makes sense to go with generating a `lazy_list` that is passed to `filter` or in this more general case to have to customize the backend.

So would it be a set of functions directly analogous to those for `summarise` and `mutate`? 
What set of functions were you thinking? 
```
filter_if(.tbl, .predicate, .funs,  .or = FALSE, ...)
filter_all(.tbl, .funs, .or = FALSE, ...)
filter_at(.tbl, .cols, .funs, .or = FALSE, ... )
```
where if `.or = TRUE` then the calls are combined by `|`.

**Comment by hadley** [comment]
@lionel- might have API suggestions too. We should make sure we have a solid API before doing any more coding (and we're happy to program this if you don't have the time/motivation)

I think the general functions belong in dplyr, and the specific filter_na probably belongs in tidyr.

**Comment by lionel-** [comment]
If I understand correctly, the suffix in `filter_all()`, `filter_at()` and `filter_if()` imply a column specification? If that's the case I think that's a very good idea. For modelling I often ignore NAs in a specific set of columns corresponding to model predictors, so I can explore the data based on the rows that are used in a model. The `filter_na()` variant could have select semantics for dots.

Maybe the `.or` thing should be a parameter in `funs()`? Or a variant of the `funs()` function might be even better, and could also be used for the predicate in other `_if()` verbs. Maybe an `ifs()` function as suggested here: https://github.com/hadley/dplyr/issues/1977, paired with an `ifs_any()`.

**Comment by hadley** [comment]
I think `.or` is out of scope here - it's not supported by `filter()` so it shouldn't be available in a filter helper. 

I'm going to close this PR in favour a more general issue on `filter_if()` etc. To achieve what you want, you'll eventually be able to just do `filter_all(data, ~ !is.na(.))`, and then that could potentially become a helper in tidyr.

**Comment by jrnold** [comment]
The idea behind `.or` was that in a manual `filter` someone is already writing it, so it's not much extra work to combine the expressions with a binary operator,
```r
filter(data, is.na(foo) | bar > 0 | baz == 1)
```
But since `filter_if` and the other filter helpers will be generating multiple expressions actually having two optional functions would be useful: a binary operator to combine expressions, and a function that takes all the other expressions as arguments. 
Getting a row with at least one missing value either requires an or operator
```
filter(data, !is.na(foo) | !is.na(bar))
```
or a function applied to the results of all the expressions
```
filter(!(is.na(foo) & is.na(bar))
```
So it may make sense to actually have to additional arguments for the filter helpers: `.c` for a binary operator used to combine the expressions and `.f` for a function applied to all of them: 
```
filter_all(data, ~ is.na(.), .c = `|`)
```
or 
```
filter_all(data, ~ is.na(.), .c = `&`, .f = `!`)
```

Maybe that isn't fully thought out, but that was my thought process when suggesting an `.or` argument for the `filter` helpers, even though it isn't necessary for `filter()`.
